### ISSUE TITLE
What (and Why) is This "Par" Pattern?

### ISSUE THREAD

**Issue by amirroth** [NREL/EnergyPlus#9110]
What is this pattern of packing a bunch of arguments into a `std::array`? I understand that functions with many arguments are bad, but this feels like the wrong way to reduce the number of arguments to a function. Are we trying to accomplish something else? If not, code clarity and maintainability would be enhanced by using normal parameters rather than this packing method. Here is an example of what I mean.

```cpp
Real64 UnitarySys::genericDXCoilResidual(
                                      EnergyPlusData &state,
                                      Real64 const PartLoadRatio,      // compressor cycling ratio (1.0 is continuous, 0.0 is off)                                                                                                                                      
                                      std::array<Real64, 8> const &Par // par(1) = DX coil number                                 
                                      )
{
	int CoilIndex = int(Par[0]);
        bool useDehumMode = int(Par[2]) > 0;
        int FanOpMode = int(Par[3]);
	int CoolingSpeedNum = int(Par[4]);
        Real64 CoolingSpeedRatio = Par[5];
        bool RunOnSensible = (Par[6] == 0.0);
        bool const singleMode = (Par[7] == 1.0);
    
   ....
    }
```

**Comment by Myoldmopar** [comment]
It was purely a way to pass a lambda and a payload into a solver function.  A root finder function takes a function reference and a payload.  Component models then ask the solver to find the root given a specific set of arguments.  In Fortran the most expedient way to get a general purpose payload was to cast everything to an array of floats and then reinterpret them once you are back inside.  With C++ there are obviously better ways.  I think with my recent SolveRoot cleanup I made it to where the payload could be anything...a tuple, a struct instance, an array.  But I can't recall for sure and I'm trying to type this fast enough that I get to be the first comment on it.

**Comment by Myoldmopar** [comment]
Yep, it's general purpose: https://github.com/NREL/EnergyPlus/blob/5bd231be385e3f90dac9a801ef30d70f8ce54f50/src/EnergyPlus/General.hh#L84

As long as your payload matches the payload type you declare as an argument to your residual function then it should work.  Now in the code right now, everything still uses the old Fortran way of creating and passes Par arrays, but those can be cleaned up easily.

**Comment by amirroth** [comment]
Can't you bind the `Par` arguments into the residual function before passing it to the solver? So that you are only passing a function with one argument (the domain) or maybe two (if you also need `state`)? That way the solver also becomes a normal function rather than a function template.

This is really the ideal application for the "lambda" constrct which allows you to bind variables from the enclosing (function) scope, i.e., reference those variables within the body of the lambda without having to pass those variables as parameters. I think you can also do this with `std::bind`, but I think that is a heavier mechanism (always use a language feature over an equivalent template feature). 

I added a short tutorial on lambdas [here](https://github.com/NREL/EnergyPlus/wiki/Lambdas:-When-and-How-to-Use-Them).
